### ISSUE TITLE
USML Visual Studio 2010 Compilation Errors

### ISSUE THREAD

**Issue by campreilly** [campreilly/UnderSeaModelingLibrary#224]
From: Coulombe, Steven P (IS) [mailto:steven.coulombe@ngc.com] 
Sent: Thursday, June 11, 2015 1:57 PM
To: Sean Reilly
Cc: Foster, James T (IS)
Subject: USML Visual Studio 2010 Compilation Errors

Sean,

We pulled down the latest USML (as of June 3rd) from the GitHub and it all compiles fine with Visual Studio 2012.  However, we have dependencies that require us to use Visual Studio 2010 for the bulk of our programs.  There are several compilation errors that occur when using VS2010 including:
```
Error C2668 ‘usml::ublas::floor’ : ambiguous call to overloaded function    seq_augment.h:64
```
I was able to do some tweaking to get past that one in the libraries only to run in to a bunch of these in compiling our program that uses the model:
```
Error C2236: unexpected token ‘struct’.  Did you forget a ‘;’?  eigenverb_notifier.h:35
Error C2332: ‘struct’ : missing tag name  eigenverb_notifier.h:35
```
These are complaining about the use of the reserved word “interface” as a variable name.

I am partly sending this to alert you to the issue but also partly to ask whether you have tested building the new stuff (post v1.0) with VS2010.

Let me know if you want additional information.

Thanks,

Steven Coulombe
Northrop Grumman Information Systems
88 Silva Lane Middletown, RI 02842

**Comment by campreilly** [comment]
Seems to refer to the master branch after merging pull request #221 from freq_comp branch.  That is commit 36d6ff511d2ed4827f628cab7a4e1595143acb7b.

**Comment by campreilly** [comment]
From: Sean Reilly 
Sent: Thursday, June 11, 2015 2:30 PM
To: 'Coulombe, Steven P (IS)'
Cc: Foster, James T (IS)
Subject: RE: USML Visual Studio 2010 Compilation Errors

Darn it.  Without really thinking it through, we all updated to VS2012.  But, we never intended to drop support for VS2010.  I’ll go ahead and revert my system to keep us in the VS2010 game.  In the mean time,

* If you know how we can fix these, feel free to fork our repository and send us a pull request.  That’s the easiest way for us to get your changes into the baseline.
* Feel free to post issues like this on our GitHub page.  I want to encourage openness in this project.

Sean

**Comment by campreilly** [comment]
From: Coulombe, Steven P (IS) [mailto:steven.coulombe@ngc.com] 
Sent: Friday, June 12, 2015 1:03 PM
To: Sean Reilly
Cc: Foster, James T (IS)
Subject: RE: USML Visual Studio 2010 Compilation Errors

Sean,

I was able to pretty easily modify the code to compile.  The first error just required some static casting – a few functions were expecting float/double types but were being passed whole number types (floor, sqrt, log10).  The second error just required the variable name change from “instance” to “instance_num”.  I did not do any testing yet to confirm that the data and calculations were not affected.  I doubt they would be but I will test before I attempt to mod the baseline. 

I have attached the changes that I made to get it to compile.  I probably won’t get around to testing these with data until sometime later next week or the week after.  If you’d rather make the changes yourself, feel free.

Thanks,

Steven Coulombe
Northrop Grumman Information Systems
88 Silva Lane Middletown, RI 02842

**Comment by campreilly** [comment]
Changes to usml for compilation w/VS2010

Changed each instance of “interface” to “interface_num” in the following files:
```
eigenverb_collection.h
eigenverb_collection.cc
eigenverb_notifier.h
eigenverb_notifier.cc
eigenverb_generator.h
eigenverb_generator.cc
eigenverb_listener.h
envelope_generator.h
envelope_generator.cc
```
Added a (double) cast as input to sqrt
```
eigenverb_test.cc:300
```
Added a static_cast<long double> as input to floor
```
seq_augment.h:64
```
Added a static_cast<long double> as input to log10
```
beam_pattern_line.cc:95
```
Added a static_cast<double> as input to sqrt
```
beam_pattern_test.cc:507
```
Changed size_t to int:
```
quadtree_test.cc:35, 37
```

**Comment by campreilly** [comment]
From: Sean Reilly 
Sent: Tuesday, June 16, 2015 10:22 AM
To: David Thibaudeau
Subject: Need help with data_grid_functors

Summary: What role is the “s” parameter supposed to play in initialize<T>::zero?

I’m working to resolve VS2010 errors that Northrup Grumman is having, and I ran into another issue.  (see https://github.com/campreilly/UnderSeaModelingLibrary/issues/224 )

With pedantic turned on, VS2010 complains that “dy3” is uninitialized at the top of data_grid::pchip()

data_grid.h:695-703
```
        DATA_TYPE pchip(int dim, const size_t* index, const double* location,
                DATA_TYPE& deriv, DATA_TYPE* deriv_vec) const
        {
            DATA_TYPE result ;
            DATA_TYPE y0, y1, y2, y3 ;             // dim-1 values at k-1, k, k+1, k+2
            DATA_TYPE dy0, dy1, dy2, dy3 ;         // dim-1 derivs at k-1, k, k+1, k+2
            initialize<DATA_TYPE>::zero(dy0, dy1, dy2, dy3) ;     // prevent valgrind from complaining
            const size_t kmin = 1u ;                      // at endpt if k-1 < 0
            const size_t kmax = _axis[dim]->size()-3u ; // at endpt if k+2 > N-1
```

It looks as if the problem might be that it is using the wrong version of initialize<T>::zero.

data_grid_functors.h:187-224
```
/**
* Base functor sets values to zero
*/
template<typename T>
struct USML_DLLEXPORT initialize {
    typedef T  argument_type ;
    typedef std::size_t    size_type ;
    static void zero( argument_type& a1, const argument_type s )
    {
        a1 = argument_type(0) ;
    }
    static void zero( argument_type& a1, argument_type& a2,
                      const argument_type s )
    {
        a1 = argument_type(0) ;
        a2 = argument_type(0) ;
    }
    static void zero( argument_type& a1, argument_type& a2,
                      argument_type& a3, const argument_type s )
    {
        a1 = argument_type(0) ;
        a2 = argument_type(0) ;
        a3 = argument_type(0) ;
    }
    static void zero( argument_type& a1, argument_type& a2,
                      argument_type& a3, argument_type& a4,
                      const argument_type s )
    {
        a1 = argument_type(0) ;
        a2 = argument_type(0) ;
        a3 = argument_type(0) ;
        a4 = argument_type(0) ;
    }
    static void value( argument_type& a, const argument_type s, argument_type v )
    {
        a = v ;
    }
};
```

When data_grid::pchip uses the 4 parameter version, it appears using the third version of zero() instead of the 4th.  If that were the case, wouldn’t it just initialize the first three?  What role is the “s” parameter supposed to play in initialize<T>::zero?  

Sean

**Comment by campreilly** [comment]
Fixed compile errors with Visual Studio 2010 32-bit
- Added a (double) cast as input to sqrt in eigenverb_test.cc:300
- Added a static_cast<long double> as input to floor seq_augment.h:64
- Added a static_cast<long double> as input to log10 in beam_pattern_line.cc:95
- Added a static_cast<double> as input to sqrt in beam_pattern_test.cc:507
- Changed size_t to int: in quadtree_test.cc:35, 37
- Remove "s" argument from initialize<T>::zero() functions.

But now, many of the tests fail.  I strongly suspect that this is because I messed up while performing "Remove "s" argument from initialize<T>::zero() functions".

Errors:
* reflection_test/reflect_grid_test has NAN for lat/long
```
../waveq3d/test/reflection_test.cc(482): error in "reflect_grid_test": difference{1.#QNAN%} between wave.curr()->position.latitude(0,0){1.#QNAN} and 36.16925316 0619995{36.169253160619995} exceeds 0.089999999999999997%
../waveq3d/test/reflection_test.cc(483): error in "reflect_grid_test": difference{1.#QNAN%} between wave.curr()->position.longitude(0,0){1.#QNAN} and 16.012084836798909{16.012084836798909} exceeds 0.089999999999999997%
```
* datagrid_test/linear_1d_test and cubic_1d_test have significant accuracy problems
```
=== datagrid_test: cubic_1d_test ===
x       truth   nearest linear  pchip
1       7.7     7.7     7.7     0       
../types/test/datagrid_test.cc(247): error in "cubic_1d_test": difference{100%} between pchip{0} and truth{7.7000000000000002} exceeds 2%
1.25    8.28125 7.7000000000000002      8.1500000000000004      0.92031250000000009     
../types/test/datagrid_test.cc(247): error in "cubic_1d_test": difference{88.8868%} between pchip{0.92031250000000009} and truth{8.28125} exceeds 2%
1.5     8.8249999999999993      7.7000000000000002      8.60000000000000142.3562500000000002      
../types/test/datagrid_test.cc(247): error in "cubic_1d_test": difference{73.3003%} between pchip{2.3562500000000002} and truth{8.8249999999999993} exceeds 2%
1.75    9.3312500000000007      7.7000000000000002      9.05000000000000074.1203125000000007   
../types/test/datagrid_test.cc(247): error in "cubic_1d_test": difference{55.8439%} between pchip{4.1203125000000007} and truth{9.3312500000000007} exceeds 2%
2       9.8000000000000007      11.300000000000001      9.5     6.0250000000000004      
../types/test/datagrid_test.cc(247): error in "cubic_1d_test": difference{38.5204%} between pchip{6.0250000000000004} and truth{9.8000000000000007} exceeds 2%
```

**Comment by campreilly** [comment]
Blocked awaiting resolution of VS2010 complains that “dy3” is uninitialized.  Even when I try to run with pedantic turned off, the code segfaults the first time that it uses pchip.

**Comment by campreilly** [comment]
After reverting changes in data_grid, still having 6 errors in sequence_test/seq_augment_test.

=== sequence_test/seq_augment_test ===
origin: [5](6,6.7839278,7.6846584,8.7389332,10)
augment: [11](6,6.1959819,6.3919639,6.5879458,6.7839278,-1.456816e+144,-1.456816e+144,-1.456816e+144,-1.456816e+144,-1.456816e+144,10)
../types/test/sequence_test.cc(607): error in "seq_augment_test": difference{1.89575e+145%} between i{-1.4568159901474629e+144} and tmp[index++]{7.68466} exceeds 0.0001%
../types/test/sequence_test.cc(607): error in "seq_augment_test": difference{1.66704e+145%} between i{-1.4568159901474629e+144} and tmp[index++]{8.7389299999999999} exceeds 0.0001%
../types/test/sequence_test.cc(607): error in "seq_augment_test": difference{1.63047e+145%} between i{-1.4568159901474629e+144} and tmp[index++]{8.93492} exceeds 0.0001%
../types/test/sequence_test.cc(607): error in "seq_augment_test": difference{1.59548e+145%} between i{-1.4568159901474629e+144} and tmp[index++]{9.1309000000000005} exceeds 0.0001%
../types/test/sequence_test.cc(607): error in "seq_augment_test": difference{1.56195e+145%} between i{-1.4568159901474629e+144} and tmp[index++]{9.3268799999999992} exceeds 0.0001%

**Comment by campreilly** [comment]
Commit 521666983db79009cdf6e19b8189d4132c3e81b0 seems to have fixed the errors with sequence_test/seq_augment_test, but now there is an error in maps_test: sensor_test.
```
=== maps_test: sensor_test ===
******
sensor_manager: add sensor(101)
sensor_pair_manager: add sensor(101)
sensor_pair_manager:  current pairs
******
sensor_manager: add sensor(212)
sensor_pair_manager: add sensor(212)
   add_multistatic_receiver: sensor_pair(101,212)
sensor_pair_manager:  current pairs
     pair  src_rcv 101_212
******
sensor_manager: update sensor(101)
sensor_model: update_sensor(101)
unknown location(0): fatal error in "sensor_test": C:\Program Files (x86)\Microsoft Visual Studio 10.0\VC\include\xutility(2225) : Assertion failed: invalid null pointer

..\sensors\test\maps_test.cc(177): last checkpoint
```

**Comment by campreilly** [comment]
The error in update_sensor appears to be because the orientation class has no operator=(), and this appears to be a problem for VS2010.
```
/**
 * Checks to see if new position and orientation have changed enough
 * to require a new WaveQ3D run.
 */
void sensor_model::update_sensor(const wposition1& position,
		const orientation& orientation, bool force_update)
{
    { // For scope of lock
        write_lock_guard guard(_update_sensor_mutex);
        if (!force_update) {
            if (!check_thresholds(position, orientation)) {
                return;
            }
        }
        #ifdef USML_DEBUG
            cout << "sensor_model: update_sensor(" << _sensorID << ")" << endl ;
        #endif
        _position = position;
        _orient = orientation;
    }
	run_wave_generator();
}
```

**Comment by campreilly** [comment]
Discovered that the problem was not that the orientation class has no operator=().  Instead the problem appears to be that problem lies in the fact that orientation passed into sensor_model::update_sensor() has an _axis attribute with a size of zero.

Copied initializers for _heading, _pitch, _roll, and _axis from default constructor into tilt angle/direction constructor in commit fa133653649332d015ba246f81bd9159889e41be.

**Comment by campreilly** [comment]
Resolved by pull request #228.
